docs/root/intro/arch_overview/http_connection_management.rst

```diff
@@ -111,6 +111,70 @@ previously attempted priorities.
 
 Envoy can be extended with custom retry plugins similar to how custom filters can be added.
 
+.. _arch_overview_internal_redirects:
+
+Internal redirects
+--------------------------
+
+Envoy supports handling 302 redirects internally, that is capturing a 302 redirect response,
+synthesizing a new request, sending it to the upstream specified by the new route match, and
+returning the redirected response as the response of the original request.
+
+There are two modes of redirect handling, "Internal redirects", for internal or private URLs,
+general redirects, which capture all 302s. They are configured separately via the
+:ref:`internal redirect action
+<envoy_api_field_route.RouteAction.internal_redirect_action>`
+and :ref:`redirect action
+<envoy_api_field_route.RouteAction.redirect_action>` fields in 
+route configuration respectively.
+
+Internal redirect handling is triggered via the x-envoy-internal-redirect header. Any response
+with this header will be treated as an internal redirect, and if the route is not configured for
+redirects, or the redirect is invalid, it will be translated to a 500 downstream. Internal redirects
+basically fail closed: it is assumed that the Location header may contain PII, or have passed auth
+checks, and should never be returned to untrusted clients. For multi-level deployments, a upstream
+Envoy can be configured to PASS_THROUGH internal redirects to the downstream Envoy, so that via
+configuration the redirects can be fielded at either level of infrastructure.
+
+Conversely, where basic redirect handling is configured on, Envoy will do best-effort handling for
+all 302 responses which do not contain an x-envoy-internal-redirect. This handling is fail-open:
+if the preconditions are not met, the 302 will be forwarded downstream.
+
+For a redirect to be successful it must pass the following checks
+
+1. Be a 302 response
+2. Have a *location* header with a valid, fully qualified URL matching the scheme of the original request.
+3. The request must have been fully processed by Envoy.
+4. The request must not have a body
+5. The request must have not been previously redirected, as determined by the presence of an x-envoy-original-url header
```

Comment:
doesn't this mandate that the downstream has to be an Envoy? which won't work for the common case.